File: SimulatorConnectObjCDemo/SimulatorConnectObjCDemo/Authentication/AuthenticationPresenter.m
//
//  AuthenticationPresenter.m
//  SimulatorConnectObjCDemo
//
//  Created by Chad Godsey on 5/18/21.
//

#import "AuthenticationPresenter.h"
#import <AuthenticationServices/AuthenticationServices.h>

@interface AuthenticationPresenter ()

@property ASWebAuthenticationSession *authSessionAS;

@end


@implementation AuthenticationPresenter

static NSString *authUrl = @"https://fsg-api.auth.us-east-1.amazoncognito.com/login";
static NSString *scope = @"openid";
static NSString *responseType = @"token";
static NSString *redirectURLEndpoint = @"oauth2redirect";
static NSString *callbackURLScheme = @"simulatorconnect";
static NSString *clientID = @"7ni0kpg1n7esk9l3cnldktt0d1";

@synthesize delegate;
@synthesize authCode;

- (void)showSignin:(UIViewController*) viewController {
    NSString *requestUrl = [NSString stringWithFormat:@"%@?scope=%@&response_type=%@&client_id=%@&redirect_uri=%@:/%@",
                            authUrl, scope, responseType, clientID, callbackURLScheme, redirectURLEndpoint];
    self.authSessionAS = [[ASWebAuthenticationSession alloc]initWithURL:[[NSURL alloc] initWithString:requestUrl]
                                                      callbackURLScheme:callbackURLScheme
                                                      completionHandler:^(NSURL * _Nullable callbackURL, NSError * _Nullable error) {
        if (error)
        {
            NSLog([error debugDescription]);
            if (self->delegate)
            {
                [self->delegate authFailure:error];
            }
        }
        
        if (callbackURL)
        {
            self.authCode = callbackURL.absoluteString;
        }else
        {
            self.authCode = @"";
        }
        if (self->delegate)
        {
            [self->delegate authSuccess:self.authCode];
        }
    }];

    self.authSessionAS.presentationContextProvider = viewController;
    self.authSessionAS.prefersEphemeralWebBrowserSession = true;
    [self.authSessionAS start];
}

@end

File: SimulatorConnectSwiftDemo/SimulatorConnectSwiftDemo/Authentication/AuthenticationPresenter.swift
//
//  AuthenticationPresenter.swift
//  SimulatorConnectSwiftDemo
//
//  Created by Chad Godsey on 5/18/21.
//

import Foundation
import AuthenticationServices


class AuthenticationPresenter : NSObject {
    static public let shared = AuthenticationPresenter()
    
    // MARK: Constants
    static let authUrl = "https://fsg-api.auth.us-east-1.amazoncognito.com/login"
    static let scope = "openid"
    static let responseType = "token"
    static let redirectURLEndpoint = "oauth2redirect"
    static let callbackURLScheme = "simulatorconnect"
    static let clientID = "7ni0kpg1n7esk9l3cnldktt0d1"
    
    public func showSignIn(completion: @escaping((String?, Error?) -> Void))
    {
        // Use the URL and callback scheme specified by the authorization provider.
        guard let authURL = URL(string: "\(AuthenticationPresenter.authUrl)?scope=\(AuthenticationPresenter.scope)&response_type=\(AuthenticationPresenter.responseType)&client_id=\(AuthenticationPresenter.clientID)&redirect_uri=\(AuthenticationPresenter.callbackURLScheme):/\(AuthenticationPresenter.redirectURLEndpoint)") else { return }

        // Initialize the session.
        let session = ASWebAuthenticationSession(url: authURL, callbackURLScheme: AuthenticationPresenter.callbackURLScheme)
        { callbackURL, error in
            // Handle the callback.
            guard error == nil, let callbackURL = callbackURL else { return }

            // The callback URL format depends on the provider. For this example:
            //   exampleauth://auth?token=1234
            let queryItems = URLComponents(string: callbackURL.absoluteString)?.queryItems
            let code = queryItems?.filter({ $0.name == "code" }).first?.value
            print("Got code \(String(describing: code))")
            completion(code, nil)
        }
        session.presentationContextProvider = self
        session.prefersEphemeralWebBrowserSession = true
        if !session.start() {
          print("Failed to start ASWebAuthenticationSession")
        }
        
    }
}


extension AuthenticationPresenter: ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        let window = UIApplication.shared.windows.first { $0.isKeyWindow }
        return window ?? ASPresentationAnchor()
    }
}
